packages/firestore/src/util/bundle.ts

```diff
@@ -0,0 +1,215 @@
+      | ReadableStream<Uint8Array | ArrayBuffer>
+      | Uint8Array
+      | ArrayBuffer
+  ) {
+    if (
+      bundleStream instanceof Uint8Array ||
+      bundleStream instanceof ArrayBuffer
+    ) {
+      this.bundleStream = toReadableStream(bundleStream);
+    }
+    this.reader = (this.bundleStream as ReadableStream).getReader();
+  }
+
+  /**
+   * Returns the metadata of the bundle.
+   */
+  async getMetadata(): Promise<BundleMetadata> {
+    if (!this.metadata) {
+      const result = await this.nextElement();
+      if (result === null || result instanceof SizedBundleElement) {
+        throw new Error(`The first element is not metadata, it is ${result}`);
+      }
+      this.metadata = (result as BundleElement).metadata;
+    }
+
+    return this.metadata!;
+  }
+
+  /**
+   * Asynchronously iterate through all bundle elements (except bundle metadata).
+   */
+  async *elements(): AsyncIterableIterator<SizedBundleElement> {
+    let element = await this.nextElement();
+    while (element !== null) {
+      if (element instanceof SizedBundleElement) {
+        yield element;
+      } else {
+        this.metadata = element.metadata;
+      }
+      element = await this.nextElement();
+    }
+  }
+
+  // Reads from the head of internal buffer, and pulling more data from underlying stream if a complete element
+  // cannot be found, until an element(including the prefixed length and the JSON string) is found.
+  //
+  // Once a complete element is read, it is dropped from internal buffer.
+  //
+  // Returns either the bundled element, or null if we have reached the end of the stream.
+  private async nextElement(): Promise<
+    BundleElement | SizedBundleElement | null
+  > {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = parseInt(lengthString, 10);
+    if (isNaN(length)) {
+      throw new Error(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    if (!this.metadata) {
+      const element = JSON.parse(jsonString) as BundleElement;
+      return element;
+    } else {
+      return new SizedBundleElement(
+        JSON.parse(jsonString),
+        lengthBuffer.length + length
+      );
+    }
+  }
+
+  // First index of '{' from the underlying buffer.
+  private indexOfOpenBracket(): number {
+    return this.buffer.findIndex(v => v === 123);
+  }
+
+  // Reads from the beginning of the inernal buffer, until the first '{', and return
+  // the content.
+  // If reached end of the stream, returns a null.
+  private async readLength(): Promise<Uint8Array | null> {
+    let position = this.indexOfOpenBracket();
+    while (position < 0) {
```

Comment:
Do you mean we always encode length in a fixed length string, like instead of `123`, we always encode it as `00000123`?

This might be easier to implement (and less error-prone), but note when dealing with `ReadableStream` we cannot control how much bytes we read back. So we still need to pull more data from the stream if it just happens that our first read is not sufficient.

Reply:
1) I don't think we should zero pad our length. While easy to implement, it seems a little hackish. We can however assume a maximum length of a number. The maximum number could be based on document size or integer boundaries (2^53, 2^64 or 2^128 - depending on other implementations).

2) Could we build a simple wrapper class that is used by both `readLength()` and `nextElement()`? This will essentially be the same as `pullMoreDataToBuffer()`, but it would allow us to implement a very simple version of `nextElement()`:

```
function nextElement() {
  const lengthBuffer = buffer.readInt(); // https://docs.oracle.com/javase/7/docs/api/java/io/DataInputStream.html#readInt()
  const json = buffer.read(lengthBuffer);
  return JSON.parse(buffer); // TODO: Validate contents
}
```

Having written this, it occurs to me that what you have is not all that different. Maybe if we can just change `readLength()` to call `readJsonString()`, which would unify the code to poll form `pullMoreDataToBuffer()`? If you added an option to not advance the buffer your calling code could be:

```
function nextElement() {
  const lengthBuffer = buffer.readJsonString(10, advance_no);  
  const length = decode(...);
  buffer.read(String(length).length, advance_yes)
  const json = buffer.read(length, advance_yes);
}
```

Sorry for my contradicting and non-coherent rambling.